File: Python/Math-Array-String/50.py
# 66. Plus One

from typing import List

class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        n = len(digits)
    
        # Traverse the array from the last digit
        for i in range(n - 1, -1, -1):
            print(i)
            if digits[i] < 9:
                # If the digit is less than 9, increment it and return the list
                digits[i] += 1
                return digits
            # If the digit is 9, set it to 0
            digits[i] = 0
        
        # If all digits are 9, add 1 at the beginning of the list
        return [1] + digits

sol = Solution()
output = sol.plusOne([9,9,9,9])
print(output) # Output: 6

File: JavaScript/Math-Array-String/50.js
// 66. Plus One


/**
 * @param {number[]} digits
 * @return {number[]}
 */
var plusOne = function(digits) {
    // Start from the last digit
    for (let i = digits.length - 1; i >= 0; i--) {
        if (digits[i] < 9) {
            // Increment the current digit if it's less than 9
            digits[i]++;
            return digits;
        }
        // If the digit is 9, set it to 0
        digits[i] = 0;
    }
    // If all digits are 9, prepend 1 to the array
    digits.unshift(1);
    return digits;
};

// Test cases
console.log(plusOne([1, 2, 3])); // Output: [1, 2, 4]
console.log(plusOne([4, 3, 2, 1])); // Output: [4, 3, 2, 2]
console.log(plusOne([9])); // Output: [1, 0]
console.log(plusOne([9, 9, 9])); // Output: [1, 0, 0, 0]
